### PR TITLE
[writeset-generator] Create the writeset patch for diem-framework-1.2.0-rc0 release

### PR DESCRIPTION
**Pull request by runtian-zhou**
## Motivation
This PR now has three commits:
1. `writeset-generator` now calls into `diem-framework`'s release api to fetch the modules to be released.
2. Creates the `release-1.2.0-rc0` writeset blob as well as artifacts needed to replay the commit.
3. This PR tested the upgrade path in e2e tests. Golden files are upgraded because of the shift in sequence number when handling upgrade path slightly differently. 

The writeset is created by following steps:
1. Checkout to `release-1.2` branch and execute `cargo run -- --output release/artifacts/release-1.2.0-rc0` in `diem-framework` module. This creates the compiled version of all Diem Framework modules on that branch top.
2. Execute the following command:
```
cargo run -- --output-payload -o release/release-1.2.0-rc0.blob create-release MAINNET https://mainnet.diem.com 15408203 --first-release --diem-version 2
```

### Have you read the [Contributing Guidelines on pull requests](https://github.com/diem/diem/blob/main/CONTRIBUTING.md#pull-requests)?

Yes

## Test Plan
Following tests are being done:
- Updated the e2e tests to use the writeset blob to test the versioning flow:
   The golden files for v2 are modified because the sequence number off by 1.
- Ran a one shot cluster test to make sure the latest binary would work with the writeset properly in #8152. The gotcha I found is we need to increment the size limit for a transaction as the writeset blob(~51kb) exceeds the default size limit set by jsonrpc (32kb).
- Use `transaction-replay` tooling to download the data in production and use `move doctor` to check if those data could be linked against the new diem framework after applying the writeset blob properly.
- Here's the modules that gets modified:
```
Adding new module: ModuleId { address: 00000000000000000000000000000001, name: Identifier("AccountAdministrationScripts") }
Adding new module: ModuleId { address: 00000000000000000000000000000001, name: Identifier("AccountCreationScripts") }
Updating existing module: ModuleId { address: 00000000000000000000000000000001, name: Identifier("AccountFreezing") }
Updating existing module: ModuleId { address: 00000000000000000000000000000001, name: Identifier("AccountLimits") }
Updating existing module: ModuleId { address: 00000000000000000000000000000001, name: Identifier("Authenticator") }
Updating existing module: ModuleId { address: 00000000000000000000000000000001, name: Identifier("BCS") }
Updating existing module: ModuleId { address: 00000000000000000000000000000001, name: Identifier("ChainId") }
Updating existing module: ModuleId { address: 00000000000000000000000000000001, name: Identifier("CoreAddresses") }
Updating existing module: ModuleId { address: 00000000000000000000000000000001, name: Identifier("DesignatedDealer") }
Updating existing module: ModuleId { address: 00000000000000000000000000000001, name: Identifier("Diem") }
Updating existing module: ModuleId { address: 00000000000000000000000000000001, name: Identifier("DiemAccount") }
Updating existing module: ModuleId { address: 00000000000000000000000000000001, name: Identifier("DiemBlock") }
Updating existing module: ModuleId { address: 00000000000000000000000000000001, name: Identifier("DiemConfig") }
Adding new module: ModuleId { address: 00000000000000000000000000000001, name: Identifier("DiemConsensusConfig") }
Updating existing module: ModuleId { address: 00000000000000000000000000000001, name: Identifier("DiemSystem") }
Updating existing module: ModuleId { address: 00000000000000000000000000000001, name: Identifier("DiemTimestamp") }
Updating existing module: ModuleId { address: 00000000000000000000000000000001, name: Identifier("DiemTransactionPublishingOption") }
Updating existing module: ModuleId { address: 00000000000000000000000000000001, name: Identifier("DiemVMConfig") }
Updating existing module: ModuleId { address: 00000000000000000000000000000001, name: Identifier("DiemVersion") }
Updating existing module: ModuleId { address: 00000000000000000000000000000001, name: Identifier("DualAttestation") }
Updating existing module: ModuleId { address: 00000000000000000000000000000001, name: Identifier("Errors") }
Updating existing module: ModuleId { address: 00000000000000000000000000000001, name: Identifier("Event") }
Updating existing module: ModuleId { address: 00000000000000000000000000000001, name: Identifier("FixedPoint32") }
Updating existing module: ModuleId { address: 00000000000000000000000000000001, name: Identifier("Hash") }
Updating existing module: ModuleId { address: 00000000000000000000000000000001, name: Identifier("Option") }
Adding new module: ModuleId { address: 00000000000000000000000000000001, name: Identifier("PaymentScripts") }
Updating existing module: ModuleId { address: 00000000000000000000000000000001, name: Identifier("RecoveryAddress") }
Updating existing module: ModuleId { address: 00000000000000000000000000000001, name: Identifier("RegisteredCurrencies") }
Updating existing module: ModuleId { address: 00000000000000000000000000000001, name: Identifier("Roles") }
Updating existing module: ModuleId { address: 00000000000000000000000000000001, name: Identifier("SharedEd25519PublicKey") }
Updating existing module: ModuleId { address: 00000000000000000000000000000001, name: Identifier("Signature") }
Updating existing module: ModuleId { address: 00000000000000000000000000000001, name: Identifier("Signer") }
Updating existing module: ModuleId { address: 00000000000000000000000000000001, name: Identifier("SlidingNonce") }
Adding new module: ModuleId { address: 00000000000000000000000000000001, name: Identifier("SystemAdministrationScripts") }
Updating existing module: ModuleId { address: 00000000000000000000000000000001, name: Identifier("TransactionFee") }
Adding new module: ModuleId { address: 00000000000000000000000000000001, name: Identifier("TreasuryComplianceScripts") }
Updating existing module: ModuleId { address: 00000000000000000000000000000001, name: Identifier("VASP") }
Adding new module: ModuleId { address: 00000000000000000000000000000001, name: Identifier("ValidatorAdministrationScripts") }
Updating existing module: ModuleId { address: 00000000000000000000000000000001, name: Identifier("ValidatorConfig") }
Updating existing module: ModuleId { address: 00000000000000000000000000000001, name: Identifier("ValidatorOperatorConfig") }
Updating existing module: ModuleId { address: 00000000000000000000000000000001, name: Identifier("Vector") }
Updating existing module: ModuleId { address: 00000000000000000000000000000001, name: Identifier("XDX") }
Updating existing module: ModuleId { address: 00000000000000000000000000000001, name: Identifier("XUS") }
```

### CODE REVIEW

**Comment by sausagee** [comment]
@runtian-zhou can you share the next steps (ie testing and rollout) for the framework release?

**Comment by runtian-zhou** [comment]
> @runtian-zhou can you share the next steps (ie testing and rollout) for the framework release?

Yea sure! The writeset should be ready to go and should contain the exact bytes we need to upgrade the diem-framework v1.2.0. I'm still discussing with the team whether the current way of generating this writeset is the way we want. Once we agreed on the generation process I'll update the oncall runbook.

**Comment by runtian-zhou** [comment]
Just renamed the writeset to `release-1.2.0-rc0` to demonstrate that it is a potential release candidate right now. We still misses a cluster test for running this blob on the release branch, which I will report shortly in #8152. Once that test passes, I'll go land this PR and use this writeset blob for testing in testnet, etc. After the testing completes we will just use this blob to upgrade the mainnet.

**Comment by runtian-zhou** [comment]
Just ran a cluster test to make sure we can use this writeset to upgrade the diem framework and enable the new feature as desired in #8152. I used the following command:
```
./scripts/cti --tag release-1.2_59d95143 --cluster-test-tag dev_runtianz_pull_8152 -E UPDATE_TO_TAG=release-1.2_b3f39961  -E COUNT=15 --suite versioning
```
And the test was executed sucessfully. However it is a bit unfortunate that we cannot land this test into release branch as a land blocking test, as I had to backport the `release-1.1` genesis code to the `release-1.2` branch in order to start a network from the old genesis. In the long term this might be fixed by being able to specify which `genesis-tool` the cluster test would like to use.

Another gotcha I found while running the cluster test is that the writeset size exceeds the maximum transaction size limit and thus couldn't be submitted at the beginning. We will need to change that config in production in order for this writeset to be submitted via jsonrpc.

**Comment by sausagee** [comment]
> Another gotcha I found while running the cluster test is that the writeset size exceeds the maximum transaction size limit and thus couldn't be submitted at the beginning. We will need to change that config in production in order for this writeset to be submitted via jsonrpc.

This is a json_rpc side config change right?  

cc @mgorven heads up on VFN config change.

**Comment by runtian-zhou** [comment]
> > Another gotcha I found while running the cluster test is that the writeset size exceeds the maximum transaction size limit and thus couldn't be submitted at the beginning. We will need to change that config in production in order for this writeset to be submitted via jsonrpc.
> 
> This is a json_rpc side config change right?
> 
> cc @mgorven heads up on VFN config change.

That is correct! It's the `content_length_limit` field

File: language/diem-tools/writeset-transaction-generator/src/main.rs
Comment (tzakian) [inline language/diem-tools/writeset-transaction-generator/src/main.rs:101]: Do we need to worry about the Debug module or other test-only modules from the move-stdlib being picked up here?

File: language/diem-tools/writeset-transaction-generator/src/main.rs
Comment (tzakian) [inline language/diem-tools/writeset-transaction-generator/src/main.rs:136]: I guess this comes back to the question above as well, but just want to make sure we're releasing the Move stdlib modules at the same time as the new diem framework modules

File: language/testing-infra/e2e-tests/src/versioning.rs
Comment (tzakian) [inline language/testing-infra/e2e-tests/src/versioning.rs:41]: ❤️

File: language/testing-infra/e2e-tests/src/versioning.rs
Comment (tzakian) [inline language/testing-infra/e2e-tests/src/versioning.rs:44]: I think it should be fine to not increment the sequence number here, but not sure if there were other issues that you encountered if you didn't bump it?

**Review by tzakian — COMMENTED** [review]
Overall LGTM. Main questions I have are on the writeset generation side and what modules are included.

File: language/testing-infra/e2e-tests/src/versioning.rs
Comment (runtian-zhou) [reply language/testing-infra/e2e-tests/src/versioning.rs:44]: Since this transaction is signed by Diem Root account I suppose its sequence number get bumped by this upgrade trasction?

File: language/diem-tools/writeset-transaction-generator/src/main.rs
Comment (runtian-zhou) [reply language/diem-tools/writeset-transaction-generator/src/main.rs:101]: We do need to worry about this IMO. There isn't a way to filter out those modules effectively. From my read is debug module isn't included in move-stdlib so it won't be there.

File: language/diem-tools/writeset-transaction-generator/src/main.rs
Comment (runtian-zhou) [reply language/diem-tools/writeset-transaction-generator/src/main.rs:136]: I guess this is a question for @vgao1996 ? To me the Move stdlib are changed a lot less frequently so it's not as a big issue as the diem framework modules.

**Review by tzakian — APPROVED** [review]
LGTM. Thanks for adding the list of changed/added modules!

**Comment by runtian-zhou** [comment]
Created tag https://github.com/diem/diem/releases/tag/diem-framework-v1.2.0 for the commit that generates the move modules

**Pull request closed without merging**
